In this practice code, we are going to use Ozone dataset. Dictionary of this dataset can be found in https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/airquality.html

In [1]:
import pandas as pd
import numpy as np

In [2]:
url = "https://raw.githubusercontent.com/ga-students/SF-DAT-20/master/Data/ozone.csv"
OzoneData = pd.read_csv(url)

#### Explore the dataset and decide which variables suffer from missing data

In [3]:
print len(OzoneData)
print OzoneData.isnull().sum()
OzoneData.describe()

153
Ozone      37
Solar.R     7
Wind        0
Temp        0
Month       0
Day         0
dtype: int64


Ozone     Solar.R        Wind        Temp       Month         Day
count  116.000000  146.000000  153.000000  153.000000  153.000000  153.000000
mean    42.129310  185.931507    9.957516   77.882353    6.993464   15.803922
std     32.987885   90.058422    3.523001    9.465270    1.416522    8.864520
min      1.000000    7.000000    1.700000   56.000000    5.000000    1.000000
25%     18.000000  115.750000    7.400000   72.000000    6.000000    8.000000
50%     31.500000  205.000000    9.700000   79.000000    7.000000   16.000000
75%     63.250000  258.750000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

Answer: Two columns, Ozone and Solar.R, have missing values

#### Let's drop rows that have missing values in all the columns you indentified above

Hint: in dropna() if you set how = 'all', it will only drop columns that are suffering from missing values at all variables you introduce in subset. If you want to get rid of the row that contains missing values in any of the variables you specify, then you shall set how = 'any'

df.dropna(how = 'all',subset = ['Var1','Var2','Var3'],inplace = True)

The above code will check if all 3 variables specified in df have missing values, if they all have missing values it will drop that row.

In [4]:
OzoneData.dropna(how='all',subset=['Ozone','Solar.R'],inplace=True)

In [5]:
len(OzoneData)

151

#### Spoiler! If everything is going according to plan you should have 151 observations remaining. Also, it seemed like the first two variables had missing values. Now please make a copy of your dataframe into a dataframe named OzoneImputeMean. Also, please use mean of the variables to fill in missing values in OzoneImputeMean

#### Note: When using dummy variables in a model and considering p-values, either keep or remove all dummy variables (do not partially remove dummy vars)

In [6]:
OzoneImputeMean = OzoneData.copy()

In [7]:
OzoneImputeMean.mean()

Ozone       42.129310
Solar.R    185.931507
Wind         9.941722
Temp        78.165563
Month        7.019868
Day         15.801325
dtype: float64

In [8]:
OzoneImputeMean.fillna(OzoneImputeMean.mean(),inplace=True)

Ozone     Solar.R  Wind  Temp  Month  Day
0     41.00000  190.000000   7.4    67      5    1
1     36.00000  118.000000   8.0    72      5    2
2     12.00000  149.000000  12.6    74      5    3
3     18.00000  313.000000  11.5    62      5    4
5     28.00000  185.931507  14.9    66      5    6
6     23.00000  299.000000   8.6    65      5    7
7     19.00000   99.000000  13.8    59      5    8
8      8.00000   19.000000  20.1    61      5    9
9     42.12931  194.000000   8.6    69      5   10
10     7.00000  185.931507   6.9    74      5   11
11    16.00000  256.000000   9.7    69      5   12
12    11.00000  290.000000   9.2    66      5   13
13    14.00000  274.000000  10.9    68      5   14
14    18.00000   65.000000  13.2    58      5   15
15    14.00000  334.000000  11.5    64      5   16
16    34.00000  307.000000  12.0    66      5   17
17     6.00000   78.000000  18.4    57      5   18
18    30.00000  322.000000  11.5    68      5   19
19    11.00000   44.000000   9.7    62      5   20
20     1.00000    8.000000   9.7    59      5   21
21    11.00000  320.000000  16.6    73      5   22
22     4.00000   25.000000   9.7    61      5   23
23    32.00000   92.000000  12.0    61      5   24
24    42.12931   66.000000  16.6    57      5   25
25    42.12931  266.000000  14.9    58      5   26
27    23.00000   13.000000  12.0    67      5   28
28    45.00000  252.000000  14.9    81      5   29
29   115.00000  223.000000   5.7    79      5   30
30    37.00000  279.000000   7.4    76      5   31
31    42.12931  286.000000   8.6    78      6    1
..         ...         ...   ...   ...    ...  ...
123   96.00000  167.000000   6.9    91      9    1
124   78.00000  197.000000   5.1    92      9    2
125   73.00000  183.000000   2.8    93      9    3
126   91.00000  189.000000   4.6    93      9    4
127   47.00000   95.000000   7.4    87      9    5
128   32.00000   92.000000  15.5    84      9    6
129   20.00000  252.000000  10.9    80      9    7
130   23.00000  220.000000  10.3    78      9    8
131   21.00000  230.000000  10.9    75      9    9
132   24.00000  259.000000   9.7    73      9   10
133   44.00000  236.000000  14.9    81      9   11
134   21.00000  259.000000  15.5    76      9   12
135   28.00000  238.000000   6.3    77      9   13
136    9.00000   24.000000  10.9    71      9   14
137   13.00000  112.000000  11.5    71      9   15
138   46.00000  237.000000   6.9    78      9   16
139   18.00000  224.000000  13.8    67      9   17
140   13.00000   27.000000  10.3    76      9   18
141   24.00000  238.000000  10.3    68      9   19
142   16.00000  201.000000   8.0    82      9   20
143   13.00000  238.000000  12.6    64      9   21
144   23.00000   14.000000   9.2    71      9   22
145   36.00000  139.000000  10.3    81      9   23
146    7.00000   49.000000  10.3    69      9   24
147   14.00000   20.000000  16.6    63      9   25
148   30.00000  193.000000   6.9    70      9   26
149   42.12931  145.000000  13.2    77      9   27
150   14.00000  191.000000  14.3    75      9   28
151   18.00000  131.000000   8.0    76      9   29
152   20.00000  223.000000  11.5    68      9   30

[151 rows x 6 columns]

In [9]:
#Check if missing values are filled
print OzoneImputeMean.isnull().sum()

Ozone      0
Solar.R    0
Wind       0
Temp       0
Month      0
Day        0
dtype: int64


#### Now it's time for imputing using linear regression lines

In [10]:
#### Before we start let's define dummy variables for variable Month - don't worry about day!
MonthDummy = pd.get_dummies(OzoneData.Month, prefix = 'Month')
del MonthDummy['Month_9']
OzoneData = pd.concat([OzoneData, MonthDummy], axis=1)
OzoneData.head()

Ozone  Solar.R  Wind  Temp  Month  Day  Month_5  Month_6  Month_7  Month_8
0     41      190   7.4    67      5    1        1        0        0        0
1     36      118   8.0    72      5    2        1        0        0        0
2     12      149  12.6    74      5    3        1        0        0        0
3     18      313  11.5    62      5    4        1        0        0        0
5     28      NaN  14.9    66      5    6        1        0        0        0

In [11]:
#now let's explore correlation Matrix
OzoneData.corr()

Ozone   Solar.R      Wind      Temp     Month       Day   Month_5  \
Ozone    1.000000  0.348342 -0.601547  0.698360  0.164519 -0.013226 -0.302963   
Solar.R  0.348342  1.000000 -0.056792  0.275840 -0.075301 -0.150275 -0.024601   
Wind    -0.601547 -0.056792  1.000000 -0.466032 -0.175317  0.042365  0.237781   
Temp     0.698360  0.275840 -0.466032  1.000000  0.397427 -0.136876 -0.637816   
Month    0.164519 -0.075301 -0.175317  0.397427  1.000000 -0.007727 -0.702257   
Day     -0.013226 -0.150275  0.042365 -0.136876 -0.007727  1.000000  0.011003   
Month_5 -0.302963 -0.024601  0.237781 -0.637816 -0.702257  0.011003  1.000000   
Month_6 -0.112006  0.023998  0.046054  0.050753 -0.362131 -0.017044 -0.242766   
Month_7  0.277961  0.176744 -0.144638  0.318103 -0.007201  0.011471 -0.247805   
Month_8  0.291808 -0.076390 -0.166105  0.321680  0.355246  0.011471 -0.247805   

          Month_6   Month_7   Month_8  
Ozone   -0.112006  0.277961  0.291808  
Solar.R  0.023998  0.176744 -0.076390  
Wind     0.046054 -0.144638 -0.166105  
Temp     0.050753  0.318103  0.321680  
Month   -0.362131 -0.007201  0.355246  
Day     -0.017044  0.011471  0.011471  
Month_5 -0.242766 -0.247805 -0.247805  
Month_6  1.000000 -0.253080 -0.253080  
Month_7 -0.253080  1.000000 -0.258333  
Month_8 -0.253080 -0.258333  1.000000

#### What seems to be the list of best variables that can define Ozone? How about Solar.R?

Answer: The variables that best define (are highly correlated with) Ozone are Temp, Wind, and Solar.R. For Solar.R, the variables that best define are Ozone and Temp.

In [12]:
# if in a multi class dummy variable case, you see a few of the dummy variables are not significant but the rest are; 
# you shall either drop all or keep all. Otherwise, selecting the base dummy will become important
from sklearn.linear_model import LinearRegression
from sklearn import feature_selection

In [13]:
# now let's use a regression model to predict Ozone. First drop NaN values in Ozone and save it in OzoneDroppedValues_Ozone
OzoneDroppedValues_Ozone = OzoneData.dropna(subset=['Ozone']).copy()

In [14]:
# then run a regression line on variables of interest and check significancy of your model
# use these variables ['Solar.R','Wind','Temp','Month_5','Month_6','Month_7','Month_8'] to predict
lm = LinearRegression()
X = OzoneDroppedValues_Ozone[['Wind','Temp','Month_5','Month_6','Month_7','Month_8']] #Michael said to not use Solar.R for this prediction
y = OzoneDroppedValues_Ozone['Ozone']

In [15]:
lm.fit(X,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [16]:
from sklearn import feature_selection
pvals = feature_selection.f_regression(X,y)[1]
print zip(X.columns.values,pvals) 
#Only Month 6 does not have a significant p-value, but it cannot be dropped because it is a dummy variable

[('Wind', 9.271973903938185e-13), ('Temp', 2.9318965924778823e-18), ('Month_5', 0.0009473179683797473), ('Month_6', 0.23128827750319397), ('Month_7', 0.0025187357056135956), ('Month_8', 0.0014814156194754573)]


In [17]:
# Now fill in null values of OzoneData['Ozone'] by predicted values
ozone_predictors = OzoneData[['Wind','Temp','Month_5','Month_6','Month_7','Month_8']]
OzoneData['ozone_predict'] = lm.predict(ozone_predictors)
OzoneData['Ozone'].fillna(OzoneData['ozone_predict'],inplace=True)

In [18]:
OzoneData.isnull().sum() #Verify null values in 'Ozone' were filled

Ozone            0
Solar.R          5
Wind             0
Temp             0
Month            0
Day              0
Month_5          0
Month_6          0
Month_7          0
Month_8          0
ozone_predict    0
dtype: int64

In [19]:
# Now repeat previous steps for Solar.R variable using ['Ozone','Wind','Temp']
OzoneDroppedValues_SolarR = OzoneData.dropna(subset=['Solar.R']).copy()
X = OzoneDroppedValues_SolarR[['Ozone','Wind','Temp']]
y = OzoneDroppedValues_SolarR['Solar.R']
lm.fit(X,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [20]:
# Check p-values of your model - if a variable is not significant, drop it
pvals = feature_selection.f_regression(X,y)[1]
print zip(X.columns.values,pvals) #Wind is not significant

[('Ozone', 0.00022176102483944005), ('Wind', 0.4959552068151224), ('Temp', 0.00075177292401030658)]


In [21]:
# Drop wind and re-reun regression
X = OzoneDroppedValues_SolarR[['Ozone','Temp']]
y = OzoneDroppedValues_SolarR['Solar.R']
lm.fit(X,y)
pvals = feature_selection.f_regression(X,y)[1]
print zip(X.columns.values,pvals) #Wind is not significant

[('Ozone', 0.00022176102483944005), ('Temp', 0.00075177292401030658)]


In [22]:
# now fill in null values of OzoneData['Solar.R'] by predicted values
solar_predictors = OzoneData[['Ozone','Temp']]
OzoneData['solar_predict'] = lm.predict(solar_predictors)

#### Now check your filled data - if your predicted values are more than maximum or less than minimum, replace them by max and min

In [23]:
solar_below_min = OzoneData['solar_predict'] < OzoneData['Solar.R'].min()
solar_above_max = OzoneData['solar_predict'] > OzoneData['Solar.R'].max()
OzoneData.loc[solar_below_min,'solar_predict'] = OzoneData['Solar.R'].min()
OzoneData.loc[solar_above_max,'solar_predict'] = OzoneData['Solar.R'].max()

#### Now will fill null values after adding bounds to imputed values

In [24]:
OzoneData['Solar.R'].fillna(OzoneData['solar_predict'],inplace=True)

In [25]:
OzoneData.isnull().sum() #Verify null values were filled

Ozone            0
Solar.R          0
Wind             0
Temp             0
Month            0
Day              0
Month_5          0
Month_6          0
Month_7          0
Month_8          0
ozone_predict    0
solar_predict    0
dtype: int64

## Bonus: In original question, replace missing values of Ozone and Solar.R by Single Imputation Regression Line with Error

In [26]:
url = "https://raw.githubusercontent.com/ga-students/SF-DAT-20/master/Data/ozone.csv"
OzoneData = pd.read_csv(url)
MonthDummy = pd.get_dummies(OzoneData.Month, prefix = 'Month')
del MonthDummy['Month_9']
OzoneData = pd.concat([OzoneData, MonthDummy], axis=1)
OzoneData.head()
OzoneData.dropna(how = 'all', subset = ['Ozone', 'Solar.R'], inplace = True)

lm = LinearRegression()

# First, impute Ozone
OzoneDroppedValues_Ozone = OzoneData.dropna(subset=['Ozone']).copy()
X = OzoneDroppedValues_Ozone[['Wind','Temp','Month_5','Month_6','Month_7','Month_8']]
y = OzoneDroppedValues_Ozone['Ozone']

lm.fit(X,y)
y_hat = lm.predict(X)
standard_error_ozone = (sum((y-y_hat)**2)/(len(y)-2))**.5
print(standard_error_ozone)

20.9139543463


In [27]:
ozone_predictors = OzoneData[['Wind','Temp','Month_5','Month_6','Month_7','Month_8']]
OzoneData['ozone_predict'] = lm.predict(ozone_predictors) + np.random.normal(0,standard_error_ozone,len(OzoneData))
ozone_below_min = OzoneData['ozone_predict'] < OzoneData['Ozone'].min()
ozone_above_max = OzoneData['ozone_predict'] > OzoneData['Ozone'].max()
OzoneData.loc[ozone_below_min,'ozone_predict'] = OzoneData['Ozone'].min()
OzoneData.loc[ozone_above_max,'ozone_predict'] = OzoneData['Ozone'].max()

In [28]:
OzoneData['Ozone'].fillna(OzoneData['ozone_predict'],inplace=True)

In [29]:
OzoneDroppedValues_SolarR = OzoneData.dropna(subset=['Solar.R']).copy()
X = OzoneDroppedValues_SolarR[['Ozone','Temp']]
y = OzoneDroppedValues_SolarR['Solar.R']
lm.fit(X,y)
y_hat = lm.predict(X)
standard_error_solar = (sum((y-y_hat)**2)/(len(y)-2))**.5
print(standard_error_solar)

86.0935501861


In [30]:
solar_predictors = OzoneData[['Ozone','Temp']]
OzoneData['solar_predict'] = lm.predict(solar_predictors) + np.random.normal(0,standard_error_solar,len(OzoneData))
solar_below_min = OzoneData['solar_predict'] < OzoneData['Solar.R'].min()
solar_above_max = OzoneData['solar_predict'] > OzoneData['Solar.R'].max()
OzoneData.loc[solar_below_min,'solar_predict'] = OzoneData['Solar.R'].min()
OzoneData.loc[solar_above_max,'solar_predict'] = OzoneData['Solar.R'].max()

In [31]:
OzoneData['Solar.R'].fillna(OzoneData['solar_predict'],inplace=True)

In [32]:
OzoneData.isnull().sum() #Verify null values were filled

Ozone            0
Solar.R          0
Wind             0
Temp             0
Month            0
Day              0
Month_5          0
Month_6          0
Month_7          0
Month_8          0
ozone_predict    0
solar_predict    0
dtype: int64